d3ploy/no_inst.py

```diff
@@ -0,0 +1,308 @@
+        default=10
+    )
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.commodity_supply = defaultdict(float)
+        self.commodity_demand = defaultdict(float)
+        self.fac_supply = {}
+        CALC_METHODS['ma'] = self.moving_avg
+        CALC_METHODS['arma'] = self.predict_arma
+        CALC_METHODS['arch'] = self.predict_arch
+
+    def enter_notify(self):
+        super().enter_notify()
+        lib.TIME_SERIES_LISTENERS[self.supply_commod].append(self.extract_supply)
+        lib.TIME_SERIES_LISTENERS[self.demand_commod].append(self.extract_demand)         
+
+    def tick(self):
+        """
+        This is the tock method for the institution. Here the institution determines the difference
+        in supply and demand and makes the the decision to deploy facilities or not.     
+        """
+        time = self.context.time
+        diff, supply, demand = self.calc_diff(time-1)
+        print(diff, supply, demand)
+        if  diff < 0:
+            proto = random.choice(self.prototypes)
+            ## This is still not correct. If no facilities are present at the start of the
+            ## simulation prod_rate will still return zero. More complex fix is required.
+            if self.fac_supply[proto]:
+                prod_rate = self.fac_supply[proto]
+            else:
+                print("No facility production rate available for " + proto)                
+            number = np.ceil(-1*diff/prod_rate)
+            i = 0
+            while i < number:
+                self.context.schedule_build(self, proto)
+                i += 1
+        if self.record:
+            with open(self.demand_commod+".txt", 'a') as f:
+                f.write("Time " + str(time-1) + " Deployed " + str(len(self.children)) + 
+                                              " supply " + str(self.commodity_supply[time-1]) + 
+                                              " demand " +str(self.commodity_demand[time-1]) + "\n")    
+
+    def calc_diff(self, time):
+        """
+        This function calculates the different in supply and demand for a given facility
+        Parameters
+        ----------        
+        time : int
+            This is the time step that the difference is being calculated for.
+        Returns
+        -------
+        diff : double
+            This is the difference between supply and demand at [time]
+        supply : double
+            The calculated supply of the supply commodity at [time].
+        demand : double
+            The calculated demand of the demand commodity at [time]
+        """
+        if not self.commodity_demand:
+            self.commodity_demand[time] = self.initial_demand
+        if not self.commodity_supply:
+            self.commodity_supply[time] = self.initial_demand
+        supply_ts = np.array()        
+        while i < time:
+            supply_ts.append(self.supply_commodity[i])                 
```

Comment:
I think this is sort of inefficient since a new copy of the array is made with each append() call. It would be better to call np.concatenate on the list. Also, where is `i` defined?  If i is non-zero, you'd write something like:

```python
supply_ts = np.concatenate(self.supply_commodity[i:time])
```